### PR TITLE
Add time based segmenting

### LINKED ISSUE #3: Time based segmenting
As it is right now, the service returns two types of data, aggregated counts and ?all to get the full records back.

The problem is rooted with the aggregated counts. They, as of today, are the sum total since the beginning of time. Which, becomes less useful as time goes on. Because in 3 months from now if I have 13,000 views on a page, this doesn't help me know how my app trends (getting better or worse view counts). This is where the usefulness of ?all comes in. With that param, you can get all of the time data about your metrics. However, this is also since the beginning of time. So, in your aggregation of your trends over time, you will need to fetch _all_ records in order to calculate the time segments of the last day, weeks, months, etc.

So, the proposal is really for both endpoint styles, count and ?all. You could have them support a 
`?since={time}` and a `?before={time}` to do really basic time range filtering. 

The important benefit here is that it is done on the server side query so it doesn't affect the performance of the endpoint over time.

### PR DESCRIPTION
**Pull request by mxstbr**
Closes #3, cc @sean-roberts

### CODE REVIEW

**Comment by mxstbr** [comment]
Dang this bumps the size above 100 LoC, it's now 102 😢  need to fix that...

File: README.md [hunk 1]
```diff
@@ -50,6 +50,10 @@ If you just want to get the views for an id and don't want to increment the view
 
 If you want to get all views for all ids, set the `all` query parameter to `true` on a root request. (i.e. `/?all=true`) If you pass the `all` parameter to an id, all ids starting with that pathname will be included. E.g. `/x?all=true` will match views for `/x`, `/xyz` but not `/y`.
 
+#### Time based segmenting
+
+By passing the `?before<time>` or `?since=<time>` parameters together with `?all=true`, you can limit the views you get returned to a certain time segment. This is useful when your collection of views becomes larger.
```
Comment (sean-roberts) [inline README.md]: missed an equals in `?before<time>`

**Comment by mxstbr** [comment]
We're also missing the ability to do `before` and `since` at the same time! Need to fit that in somehow...

**Comment by sean-roberts** [comment]
I think Before/After are good too. Go with that.

The code looks fine as it is.

I do think, however, the API could be better. (going with before/after) I think that `?before` and `?after` should be supported for both endpoints types (all/count). With the understood downside that it's not as performant of a query. But I think the primary reason most people would use get time-segments is to get the counts between those times. So providing it in the GET count-based requests requires no extra work to get the details. The usefulness of ?all is primarily targeted at some form of more advanced interface like a UI or CLI to handle calculations for the user.

I also think that you should support the combination of `?after={time}&before={time}` in which should 
- return nothing if before < after
- return a filter of the metrics between the two bounds. `return m.time < before && m.time > after;`

**Comment by mxstbr** [comment]
Added support for specifying both `before` and `after`. Not sure what you mean by this being support for both endpoints? What's the use of supporting `/path?before=123`?

**Comment by mxstbr** [comment]
Should we return an error if `before` and `after` are the wrong way around, i.e. if `before < after`?

**Comment by sean-roberts** [comment]
I think an error would be fine for that.

So, I don't actually know why anyone would use `?before` by itself. But it comes pretty cheap to support if they did.

**Comment by mxstbr** [comment]
It's already supported 😉 will change to an error!

**Comment by luisrudge** [comment]
I'm not sure you should filter in the API itself. You should filter in the db.js file, so we can implement another db with filtering later.

**Comment by sean-roberts** [comment]
I tend to agree @luisrudge

**Comment by mxstbr** [comment]
I actually agree @luisrudge, but that's out of scope for this PR I think. If we do that, we also have to move the filtering of the keys to `db.js` and fix the whole testing setup.

Let's ship this for now and rework in another PR when we introduce database adapters?

**Comment by sean-roberts** [comment]
Is the PR going to support the before & after on GET?

**Comment by mxstbr** [comment]
> Is the PR going to support the before & after on GET?

It doesn't currently, what's the use? `GET /id` will always just return a single view...

**Comment by sean-roberts** [comment]
https://github.com/mxstbr/micro-analytics/pull/4#issuecomment-271626714

**Comment by mxstbr** [comment]
Yeah I didn't quite get that 😉 would you mind elaborating? Where do you want to see it supported?

**Comment by sean-roberts** [comment]
ha, I'll try again

So my thought process is, what is the benefit from using `?all`? It's to get the detailed timing of events. But I don't think people will commonly use it for the detail that it provides. I think people would use it to figure out the total views in a time segments. That is because, as time goes on, `GET + ?inc=false` becomes less useful. So people will fall back to having to do `?all` and then have to add logic to do the counts of views. 

With that, I think if `GET + ?inc=false` is used the user should be able to get the views count AND have the option to submit `before` and `after` time segments. It would just make the experience of that endpoint remain useful as time goes on.

So the result of this work would be, you can use `?before={time}`, `?after={time}, or `?before={time}&after={time}` on the `?all` endpoint and the `GET + ?inc=false` endpoint. Allowing for quick aggregated stats on the `GET + ?inc=false` that can be time scoped. And more advanced results for `?all` that can also be scoped.

Does that resonate better?

**Comment by mxstbr** [comment]
Good call, I totally forgot about `?inc=false`! I think we might need to do this as part of #8, as doing it in core now and then moving that over will be a lot harder...

**Comment by mxstbr** [comment]
Closing in favor of landing #14!

**Pull request closed without merging**